### DISCUSSION TITLE
async with session.begin(): sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.

### DISCUSSION THREAD

**Discussion by nitanmarcel** [sqlalchemy/sqlalchemy#6756] [Usage Questions]
I'm coming from the non async version of sqlalchemy and I have some problems with some usage of the async one.

Based on the old API I'm trying to fetch a table, and update the values in it or create it in case the table row doesn't exist.

The old code I would use would be something like

```
class Version(declarative_base()):
    __tablename__ = "versions"
    name = Column(String)

curr = session.query(Version).get("0.1")
if not curr:
    curr = Version("0.1")
session.add(curr)
sessioncommit()
```

Now with the new API my code looks something like this:

```
class Version(declarative_base()):
    __tablename__ = "versions"
    number = Column(Integer, primary_key=True)

async def insert_into_version(number):
    async with async_session() as session:
        query = select(VERSION)
        result = await session.execute(query)
        curr = result.scalars()
        if not curr:
            curr = Version(number=number)
        else:
            curr.number = number
        async with session.begin():
            session.add(curr)

```

But the code above ends with `sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.` and I don't really understand why since I made sure that excepting the creation of the tables I have no other transactions running.

Here's how I create all the tables and declare the session:

```
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

from .. import CONFIG, LOOP

def _build_async_db_uri(uri):
    if "+asyncpg" not in uri:
        return '+asyncpg:'.join(uri.split(":", 1))
    return uri


engine = create_async_engine(_build_async_db_uri(CONFIG.general.sqlalchemy_db_uri))

async def start():
    async with engine.begin() as conn:
        await conn.run_sync(BASE.metadata.create_all)

BASE = declarative_base()
LOOP.run_until_complete(start())
async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)
```

Without using `async with session.begin()` I end up with: `Class 'sqlalchemy.engine.result.ScalarResult' is not mapped  `

**Comment by CaselIT** [comment]
Hi,

The relevant part of the documentation is https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#session-subtransactions

**Comment by zzzeek** [comment]
> ```
> 
> async def insert_into_version(number):
>     async with async_session() as session:
>         query = select(VERSION)
>         result = await session.execute(query)
>         curr = result.scalars()
>         if not curr:
>             curr = Version(number=number)
>         else:
>             curr.number = number
>         async with session.begin():
>             session.add(curr)
> ```
> 
> But the code above ends with `sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.` and I don't really understand why since I made sure that excepting the creation of the tables I have no other transactions running.

the transaction is started when you ran "await session.execute(query)".     To run the whole block in an explicit transaction you would write it as (note I'm assuming "async_session" is a sessionmaker):

```py
async with async_session.begin() as session:
     # do things
```

> 
> Without using `async with session.begin()` I end up with: `Class 'sqlalchemy.engine.result.ScalarResult' is not mapped `

that sounds like a totally different issue, such as you are executing a query, getting a result back and calling result.scalars(), but then you aren't actually fetching the objects, and then feeding that ScalarResult object into an ORM query of some kind.   Would need to see stack trace and MCVE.

**Reply by nitanmarcel** [reply]
I played with the documentation and I managed to make everything work. And I have no idea on how I produced the respective error
